File: 2-dataclasses/address.kt
import java.time.LocalDate

data class Address(
    val country: String,
    val state: String,
    val suburb: String,
    val street: String,
    val houseNumber: Int,
    val unitNumber: Int? = null
)

data class Person(
    val name: String,
    val dob: LocalDate
)


fun main() {
    val addressbook = mutableMapOf<Address, List<Person>>()
    val address1 = Address("Australia", "WA", "Baywater", "Guildford Rd", 25)
    val person = Person("Foo Bar", LocalDate.of(2022, 3, 1))
    addressbook[address1] = listOf(person)

    val address2 = address1.copy(houseNumber=26)
    addressbook[address2] = listOf(person)

    print(addressbook)
}

File: 5-dataclasses/address.py
from dataclasses import dataclass, replace
from datetime import date
from pprint import pprint


@dataclass(frozen=True)
class Address:
    country: str
    state: str
    suburb: str
    street: str
    house_number: int
    unit_number: int | None = None


@dataclass
class Person:
    name: str
    dob: date


addressbook: dict[Address, list[Person]] = {}

address1 = Address('Australia', 'WA', 'Baywater', 'Guildford Rd', 25)
person = Person('Foo Bar', date(2022, 3, 1))
addressbook[address1] = [person]

address2 = replace(address1, house_number=26)
addressbook[address2] = [person]

pprint(addressbook)
